packages/ui/src/NumberField.tsx

```diff
@@ -0,0 +1,118 @@
+import { DataTestProp } from '@hazelcast/helpers'
+import React, { FC, FocusEvent, ChangeEvent, ReactElement, InputHTMLAttributes, useCallback, useLayoutEffect } from 'react'
+import { Plus, Minus } from 'react-feather'
+import cn from 'classnames'
+
+import { TextField } from './TextField'
+import { IconButton } from './IconButton'
+
+import styles from './NumberField.module.scss'
+
+type NumberFieldCoreProps = {
+  name: string
+  value: number
+  onBlur?: (e: FocusEvent<HTMLInputElement>) => void
+  onChange: (newValue: number) => void
+  error?: string
+}
+export type NumberFieldExtraProps = {
+  incrementIconAriaLabel?: string
+  decrementIconAriaLabel?: string
+  step?: number
+  min?: number
+  max?: number
+  numberType?: 'int' | 'float'
+  label: string
+  required?: boolean
+  helperText?: string | ReactElement
+  className?: string
+  inputClassName?: string
+  errorClassName?: string
+  placeholder: string
+} & DataTestProp &
+  Pick<InputHTMLAttributes<HTMLInputElement>, 'autoFocus' | 'disabled' | 'autoComplete'>
+type NumberFieldProps = NumberFieldCoreProps & NumberFieldExtraProps
+
+export const NumberField: FC<NumberFieldProps> = ({
+  incrementIconAriaLabel = 'Increment',
+  decrementIconAriaLabel = 'Decrement',
+  step = 1,
+  min,
+  max,
+  value,
+  numberType = 'int',
+  className,
+  onChange,
+  name,
+  ...props
+}) => {
+  useLayoutEffect(() => {
+    if (min !== undefined && value < min) {
+      onChange(min)
+    }
+    if (max !== undefined && value > max) {
+      onChange(max)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [min, max])
+
+  const onDecrement = useCallback(() => {
+    let newValue = value - step
+    if (min !== undefined && newValue < min) {
+      newValue = min
+    }
+    onChange(newValue)
+  }, [value, onChange, step, min])
+
+  const onIncrement = useCallback(() => {
+    let newValue = value + step
+    if (max !== undefined && newValue > max) {
+      newValue = max
+    }
+    onChange(newValue)
+  }, [value, onChange, step, max])
+
+  const overlay = (
```

Comment:
nit: we can memoize this.

Reply:
onIncrement and onDecrement are updated with every value change. So `overlay` is updated as well. I just feel it would be updated too frequently for `useMemo` to truly shine here. WDYT?

Reply:
Wouldn't the same logic apply to `useCallback` used for `onDecrement` and `onIncrement`? 
Anyway, you're right about frequent updates. Let's leave it as is.

Reply:
You are right. It should apply. Updated!